component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/EventHubClient.java

```diff
@@ -32,74 +45,218 @@
+                            this.clientStatus = ClientStatus.RETRYING;
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, false);
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else if (e instanceof AmqpException && ((AmqpException) e).getErrorCondition()
+                                == AmqpErrorCondition.RESOURCE_LIMIT_EXCEEDED) {
+                            //If resource limit is exceeded we will retry after a constant delay
+                            log.error("Resource limit exceeded when publishing EventDataBatch. Operation will be "
+                                              + "retried after constant delay");
+                            try {
+                                Thread.sleep(1000 * 60);
+                            } catch (InterruptedException interruptedException) {
+                                Thread.currentThread().interrupt();
+                            }
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else {
+                            //For any other exception
+                            log.error("Unknown error occurred while publishing EventDataBatch. Producer client will "
+                                              + "be re-initialized. Events may be lost in the process.");
+                            this.clientStatus = ClientStatus.RETRYING;
+                            readWriteLock.writeLock().unlock();
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, true);
+                        }
+                    } finally {
+                        try {
+                            readWriteLock.writeLock().unlock();
+                        } catch (IllegalMonitorStateException e) {
+                            //ignore
+                        }
                     }
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                } finally {
-                    sendSemaphore.release();
+                }
+                if (isAdded) {
+                    log.debug("Adding event: " + event.replaceAll("[\r\n]", ""));
+                } else {
+                    log.debug("Failed to add event: " + event.replaceAll("[\r\n]", ""));
+                }
+            } finally {
+                try {
+                    readWriteLock.readLock().unlock();
+                } catch (IllegalMonitorStateException e) {
+                    //ignore
                 }
             }
+        } else {
+            try {
+                threadBarrier.lock();
+                log.debug(Thread.currentThread().getName().replaceAll("[\r\n]", "") + " will be parked as EventHub "
+                                  + "Client is inactive.");
+                waitCondition.await();
```

Comment:
Shouldn't we only execute this part if the ClientStatus is RETRYING?

Reply:
Publisher threads will come to this point only when we encounter an intermittent error(successful startup). In such a case some error can be recoverable and some can be non-recoverable. But even in the non-recoverable case we need to park the publisher thread. Hence this block serves that purpose as well. 
In parallel is the error is non-recoverable PT threads will drop the messages at the Metric level